src/wam2layers/preprocessing/pressure_levels.py

```diff
@@ -0,0 +1,139 @@
+    da_p.attrs["units"] = "Pa"
+
+    p = da_p.to_numpy()
+    u = input["u"].to_numpy()
+    v = input["v"].to_numpy()
+    q = input["q"].to_numpy()
+
+    # Insert top of atmosphere values
+    # Assume wind at top same as values at lowest pressure, humidity at top = 0
+    p0 = p.argmin(axis=level_ax).flatten()[
+        0
+    ]  # index along level axis of minimum pressure
+    p = np.insert(p, 0, 0.0, axis=level_ax)
+    q = np.insert(q, 0, 0.0, axis=level_ax)
+    v = np.insert(v, 0, np.take(v, p0, axis=level_ax), axis=level_ax)
+    u = np.insert(u, 0, np.take(u, p0, axis=level_ax), axis=level_ax)
+
+    # Insert surface level values
+    p = np.insert(p, 0, surface["ps"], axis=level_ax)
+    q = np.insert(q, 0, surface["qs"], axis=level_ax)
+    v = np.insert(v, 0, surface["vs"], axis=level_ax)
+    u = np.insert(u, 0, surface["us"], axis=level_ax)
+
+    # Sort arrays by pressure (ascending)
+    u = sortby_ndarray(u, p, axis=level_ax)
+    v = sortby_ndarray(v, p, axis=level_ax)
+    q = sortby_ndarray(q, p, axis=level_ax)
+    p = sortby_ndarray(p, p, axis=level_ax)
+
+    # Insert boundary level values (at a dummy pressure value)
+    p_boundary = 0.72878581 * np.array(surface["ps"]) + 7438.803223
+    u = np.insert(u, 0, interpolate(p_boundary, p, u, axis=level_ax), axis=level_ax)
+    v = np.insert(v, 0, interpolate(p_boundary, p, v, axis=level_ax), axis=level_ax)
+    q = np.insert(q, 0, interpolate(p_boundary, p, q, axis=level_ax), axis=level_ax)
+    p = np.insert(p, 0, p_boundary, axis=level_ax)
+
+    # Sort arrays by pressure once more (ascending)
+    u = sortby_ndarray(u, p, axis=level_ax)
+    v = sortby_ndarray(v, p, axis=level_ax)
+    q = sortby_ndarray(q, p, axis=level_ax)
+    p = sortby_ndarray(p, p, axis=level_ax)
+
+    # Reset level coordinate as its values have become meaningless
+    nlev = np.size(u, axis=level_ax)
+    levs = np.arange(nlev)
+
+    # reconstruct dataarrays
+    coords = dict(input["u"].coords)
+    coords["level"] = levs
+
+    output = xr.Dataset()
+    output["u"] = xr.DataArray(u, coords, input["u"].dims, attrs=input["u"].attrs)
+    output["v"] = xr.DataArray(v, coords, input["v"].dims, attrs=input["v"].attrs)
+    output["q"] = xr.DataArray(q, coords, input["q"].dims, attrs=input["q"].attrs)
+    output["p"] = xr.DataArray(p, coords, input["u"].dims, attrs=surface["ps"].attrs)
+    pb = xr.DataArray(p_boundary, surface["ps"].coords, surface["ps"].dims)
+
+    return output, pb
+
+
+def interp_dp_midpoints(
+    extended_data: xr.Dataset,
+    ps: xr.DataArray,
+):
+    """Interpolate the data to midpoints to allow for integration to two layers.
+
+    Args:
+        Dataset with the following variables:
+            q: Specific humidity at levels
+            u: Eastward horizontal wind speed at levels
+            v: Northward horizontal wind speed at levels
+            p: Air pressure at levels
+            ps: Air pressure at the surface
```

Comment:
`ps` is a separate input argument. Also explain why it is needed? This function still does two things, also masking data below surface.

Reply:
Good point. I split off the masking to a separate function